### ISSUE TITLE
logout method not clearing credentials from cache

### ISSUE THREAD

**Issue by DarkLite1** [AzureAD/microsoft-authentication-library-for-js#2123]
## Library
- [x] `@azure/msal-browser@2.0.1`

## Framework
Quasar Framework with Vue

## Description
After upgrading from version `2.0.0` to version `2.0.1` we receive the typescript error below. The strange thing is that when logging out of the app by calling the function below the credentials are still in the cache The user is indeed logged out but when he tries to log in again he is not asked for his password again and simply gets signed in again with a quick empty pop-up

```typescript
export const logout = () => {
  const logoutRequest = {
    account: auth.getAccountByUsername(account.value.username),
  }
  void auth.logout(logoutRequest)
  setAccount()
}
```

[GitHub link](https://github.com/DarkLite1/hip/blob/master/src/services/auth/authService.ts)

## Error Message
```
TS2345: Argument of type '{ account: Msal.AccountInfo | null; }' is not assignable to parameter of type 'EndSessionRequest'.
  Types of property 'account' are incompatible.
    Type 'AccountInfo | null' is not assignable to type 'AccountInfo | undefined'.
      Type 'null' is not assignable to type 'AccountInfo | undefined'.
```

## Security
- [ ] Is this issue security related?

## Regression
- [ ] Did this behavior work before? 
Version:

## MSAL Configuration

```js
// Provide configuration values here.
// For Azure B2C issues, please include your policies.
```

## Reproduction steps
```js
// Provide relevant code snippets here.
// For Azure B2C issues, please include your policies.
```

## Expected behavior
We would expect to have the credentials cleared from the cache and request for username and password again after a logout has been performed. Even when using `logout()` the credentials aren't requested again.

## Browsers/Environment
- [x] Chrome

**Comment by pkanher617** [comment]
@DarkLite1 Do you have a stack trace showing which line is throwing this error?

> The strange thing is that when logging out of the app by calling the function below the credentials are still in the cache The user is indeed logged out but when he tries to log in again he is not asked for his password again and simply gets signed in again with a quick empty pop-up

This is expected behavior. We clear the cache, but no request is sent to the server, so the session on the server still exists.

**Comment by DarkLite1** [comment]
Thank you for getting back to me.

The error is thrown from this line:
```typescript
logout({ account: auth.getAccountByUsername(account.value.username), })
// or
logout({ account:: null })
```
I'm feeding it `null` instead of `undefined` as required by type `EndSessionRequest` of `logout()` method.


> @pkanher617 : This is expected behavior. We clear the cache, but no request is sent to the server, so the session on the server still exists.


Is there a way to invalidate the session on the server when a user logs out? What is otherwise the purpose of calling `logout()` if another user can sit in front of the PC click `login` again and be logged in as the previous user....

**Comment by pkanher617** [comment]
> Is there a way to invalidate the session on the server when a user logs out? What is otherwise the purpose of calling logout() if another user can sit in front of the PC click login again and be logged in as the previous user....

Sorry I should have been more clear. There is no other way to invalidate a session in the browser other than redirecting to the logout url. That is what the `logout()` method does. Due to the error you are receiving, we are clearing the cache but the redirect is not happening. 

> The error is thrown from this line:

This isn't helpful. we need the stack trace to see what line is erroring in our library.

`EndSessionRequest` does not accept null currently. Only an object or undefined. The reason this is erroring now is because in 2.0.1 we have updated `getAccountByUsername` function return signature to be `AccountInfo|null`. I recommend changing your code to the following

```javascript
const accountObj =  auth.getAccountByUsername(account.value.username);
auth.logout(accountObj ? { account: accountObj } : null);
```

**Comment by DarkLite1** [comment]
Thank you again for getting back to me.  Some things that are still a bit unclear to me:

> When the `logout()` method is called, the user is redirected to the page where he needs to select his account. If he does select his account and continues, or does not select his account and goes back (redirectFlow non pop-up) he gets correctly signed out.

- Why present the user with a selection of accounts to logout from, if it's not needed to select one to be signed out?

> When the user successfully signed out and comes back on the page and clicks the login button again he gets signed in without any question for user name or password.

- Why is the user not asked for his password after he explicitly signed out with `logout()`? 

This seems to me like a security risk. The implementation for `signin()` and `logout()` can be found [here](https://github.com/DarkLite1/hip/blob/master/src/services/auth/authService.ts). Maybe I'm using it wrong. But I expect that once a user logs out explicitly and he wants to sign in again he should enter at the very minimum his password again. Thinking about shared PC's and such here....

- What is the difference in calling `logout()` without an account and `logout(account: accountObj)`?

Thank you for shedding some light on these questions.

**Comment by pkanher617** [comment]
@DarkLite1 The server and client manage accounts separately. When you call the `logout()` function on MSAL, we clear all account info on the client, then redirect you to the authorization server, where you are presented with a logout screen. If you only have one account logged in, this will simply inform you that you are logged out and then redirect. If the server detects multiple accounts, it will ask you which one to log out of on the server, then invalidate that account.

> Why is the user not asked for his password after he explicitly signed out with logout()?

This is probably not intended behavior. If you chose the account to logout of on the account selection screen, and you are getting silent login when you sign in again for the same account, we should understand why that is happening. I can't repro that behavior in my environment. 

> What is the difference in calling logout() without and account with an accountlogout(account: accountObj)?
In both cases I get to choose a user to logout after the redirect.

The difference is that you are choosing to clear data for all accounts on the client vs one account on the client. It has no effect on  the server session for the account, which is why you still have to choose accounts in both scenarios. We are working with the server to get a better experience for the `logout()`.

**Comment by DarkLite1** [comment]
> > Why is the user not asked for his password after he explicitly signed out with logout()?
> 
> This is probably not intended behavior. If you chose the account to logout of on the account selection screen, and you are getting silent login when you sign in again for the same account, we should understand why that is happening. I can't repro that behavior in my environment.

@pkanher617 in case fo signing out and invalidating a specific user session on the server the password for this user still seems to be remembered my msal in the local cache. Because when that users tries to sing in again (after successful logout) he clicks his account and is not asked for the password. 

This seems to me like a security risk. In my understanding msal should revoke the active session from the server, which it clearly does AND remove the credentials from the client (browser's) cache. This last step is not happening, that's why the user can simply login again without entering his password. 

I'll try to setup a small repro.

**Comment by DarkLite1** [comment]
Added repro case:

### Setup test environment
```
git clone git@github.com:DarkLite1/msal.git

// Install the dependencies
yarn

// Update Azure credentials
// Add you details here: src\app-config.json

// Start the app
quasar dev
```

### Run test

1. Click the `LOGIN` button at the top right and sign in with your account
> Notice Welcome Bob text on the home page and the login button changed to Bob

2. Click the same button again with your name on and then click `LOGOUT`, select your account
> You are now logged out correctly, see `LOGIN` button top right, no longer your name

3. Click the `LOGIN` button at the top right again and select your account
> Notice no password is asked and you are signed in immediately

IMHO, because you performed a `LOGOUT` there should be a password question upon `LOGIN`.

**Comment by DarkLite1** [comment]
@pkanher617 can you try the test case?

**Comment by DarkLite1** [comment]
Sp np feedback on this anymore?

**Comment by tnorling** [comment]
@DarkLite1 Apologies for letting this slip, I'd like to clarify a couple things and then see if you can help us determine what's going on in your case.

> in case fo signing out and invalidating a specific user session on the server the password for this user still seems to be remembered my msal in the local cache.

Passwords are not stored anywhere in the browser nor are they ever known to msal. The 2 storage related items to be aware of are: the tokens stored in local/session storage which belong to the msal context and a cookie which represents your session with AAD which belongs to the server context. 

> In my understanding msal should revoke the active session from the server, which it clearly does AND remove the credentials from the client (browser's) cache. This last step is not happening, that's why the user can simply login again without entering his password.

Can you clarify how you know that the session on the server is invalidated?

Logout is a 2 step process:
1. Clear local/session storage for the account you specified
2. Redirect to AAD where you will be prompted to select an account to log out of. Once selected the session will be invalidated and the server cookie will be removed.

If you close the browser window or an error occurs before or during step 2 you may get the behavior you mentioned because msal thinks you're logged out but the server has not invalidated the session yet. If you are able to complete the whole flow and are still experiencing this we should understand why.

So with that said can you confirm for us whether local/session storage is being cleared after invoking `logout` indicating the user has been logged out of msal? If so, can you email one of us network traces including both logout and the subsequent login so we can follow up with the server team? Thanks!

**Comment by DarkLite1** [comment]
Thank you for taking the time to go through all of this and trying to clear up some stuff for a newbie like me. Sorry about phrasing it incorrectly with storing the password. That part wasn't clear, but the issue still stands.

The two logout steps are happening correctly. When we call the `logout()` method the `sessionStorage` is being cleared completely and the user is redirected to AAD to select his account to invalidate the session on the server.

So after login we get this:
![image](https://user-images.githubusercontent.com/10828570/93301085-361c5d00-f7f8-11ea-96c6-5ef3385ab1ec.png)

And after logout SessionStorage is correctly cleared:
![image](https://user-images.githubusercontent.com/10828570/93301198-619f4780-f7f8-11ea-8bd2-b8b4650099ff.png)

And we get back to the login page:
![image](https://user-images.githubusercontent.com/10828570/93301284-872c5100-f7f8-11ea-8abe-95dd468ce208.png)

When we then call `loginPopup` we can correctly select the user:
![image](https://user-images.githubusercontent.com/10828570/93301468-cc508300-f7f8-11ea-92db-7b22f3d7b726.png)

When we do select the correct user no password is asked again, although this user's session was invalidated before:
![image](https://user-images.githubusercontent.com/10828570/93301615-ff931200-f7f8-11ea-92fb-ac4c1a6d1196.png)

I'll try to figure out how to get you a Network log...

**Comment by DarkLite1** [comment]
@tnorling I've sent you the log files by mail Hope you see something that might clear up the confusion.

**Comment by tnorling** [comment]
@DarkLite1 Based on the screenshots you shared above it seems logout is behaving as expected. The fact that you are prompted to select an account proves that the server session **for this app** was closed. You can verify this by clearing localstorage manually then refreshing the page and logging in. You'll notice the behavior is slightly different in that the server puts you right through instead of asking which account you want to log in with. Also notice on the sign out screen the words "Signed in" are shown under your user but that is no longer present on the subsequent login screen.

What I think is happening is that [Seamless SSO](https://docs.microsoft.com/en-us/azure/active-directory/hybrid/how-to-connect-sso) or some other form of SSO is signing you back in silently. I can verify this with the network logs but the network logs you sent don't include the actual sign-in action. (If you're using popups and capturing in chrome that may be why). Try using `loginRedirect` or using Fiddler to capture the network traffic.

If you really really want to force the user to enter their password every time, you can pass `prompt: "login"` to your `loginPopup` call.

**Comment by DarkLite1** [comment]
Thank you for getting back to me @tnorling . I've checked our Azure AD and it seems like Seamless SSO is Disabled in our case:
![image](https://user-images.githubusercontent.com/10828570/93428830-6e876e00-f8c0-11ea-9af4-0daacf6a8e74.png)

I've sent you the correct log files with the redirect method this time. What I would expect is:

- A users logs in, does his thing, signs out and another user can't simply click the login button to sign in again as that user. 

This is currently not working. I have also tried your suggestion by adding `prompt: "login"` to `loginPopup`. What this does is request every time the user e-mail address and when entered no password is requested anymore. So it has the same result.  

I find it strange that once a user logs out, on a subsequent login the password is no longer needed. This seems like a security risk to me. Consider a kiosk machine where a user logs in to check his mail and afterwards logs out. If another user would go there and click the login button he would be able to sign in as the previous user and read all his emails....
